lib/net.c

```diff
@@ -1200,12 +1201,102 @@ start_iface(struct gatekeeper_if *iface, unsigned int num_attempts_link_get)
 	return ret;
 }
 
+static void
+calculate_net_config_para(struct net_config *net_conf)
+{
+	int i;
+
+	/* The maximum number of RX queues on any interface. */
+	uint16_t gatekeeper_num_rx_queues = net_conf->back_iface_enabled ?
+		RTE_MAX(net_conf->front.num_rx_queues,
+			net_conf->back.num_rx_queues) :
+		net_conf->front.num_rx_queues;
+
+	/* The maximum number of TX queues on any interface. */
+	uint16_t gatekeeper_num_tx_queues = net_conf->back_iface_enabled ?
+		RTE_MAX(net_conf->front.num_tx_queues,
+			net_conf->back.num_tx_queues) :
+		net_conf->front.num_tx_queues;
+
+	const char *block_names[] = { "dynamic_conf" };
+
+	/*
+	 * The total number of lcores used by any functional block
+	 * for either RX or TX except for Dynamic Configuration.
+	 */
+	uint16_t gatekeeper_num_lcores = get_filtered_num_lcores(block_names,
+		sizeof(block_names) / sizeof(*block_names));
+
+	/*
+	 * The number of elements in the mbuf pool.
+	 *
+	 * The optimum size (in terms of memory usage) for a mempool is when
+	 * it is a power of two minus one.
+	 *
+	 * Need to provision enough memory for the worst case. According to
+	 * l3fwd sample application, it's essentially the number of RX
+	 * descriptors (across all queues for all ports), the number of TX
+	 * descriptors (across all queues for all ports), the number of packet
+	 * burst buffers (across all lcores for all bursts), the number of
+	 * slots in the packet buffer cache (across all lcores).
+	 */
+	net_conf->gatekeeper_num_mbuf = rte_align32pow2(
+		net_conf->gatekeeper_num_ports * gatekeeper_num_rx_queues *
```

Comment:
1. The product `net_conf->gatekeeper_num_ports * gatekeeper_num_rx_queues` is an approximation of the total number of RX queues, which we'll exactly obtain after another comment that explains how to update `gatekeeper_num_rx_queues`. That is, with this change, we can replace the product `net_conf->gatekeeper_num_ports * gatekeeper_num_rx_queues` with simply `gatekeeper_num_rx_queues`.

2. Apply item 1 to product `net_conf->gatekeeper_num_ports * gatekeeper_num_tx_queues` as well.

Notice that we don't even need to update the comment of `net_conf->gatekeeper_num_mbuf` after this change because we are just making the value more precise.